src/runtime.rs

```diff
@@ -280,113 +247,151 @@ impl Runtime {
-            for active_query in &mut *cycle_participants {
-                cycle.push(active_query.database_key_index);
-            }
-
-            assert!(!cycle.is_empty());
-
-            for active_query in cycle_participants {
-                active_query.cycle = cycle.clone();
-            }
-
-            crate::CycleError {
-                cycle,
-                changed_at,
-                durability: Durability::MAX,
-            }
-        } else {
-            // Part of the cycle is on another thread so we need to lock and inspect the shared
-            // state
-            let dependency_graph = self.shared_state.dependency_graph.lock();
-
-            let mut cycle = Vec::new();
-            dependency_graph.push_cycle_path(
+        to_id: RuntimeId,
+    ) -> CycleDetected {
+        debug!("create_cycle_error(database_key={:?})", database_key_index);
+
+        let mut from_stack = self.local_state.take_query_stack();
+        let from_id = self.id();
+
+        // Extract the changed_at and durability values from the top of the stack;
+        // these reflect the queries which were executed thus far and which
+        // led to the cycle.
+        let changed_at = from_stack.last().unwrap().changed_at;
+        let durability = from_stack.last().unwrap().durability;
+
+        // Identify the cycle participants:
+        let cycle = {
+            let mut v = vec![];
+            dg.for_each_cycle_participant(
+                from_id,
+                &mut from_stack,
                 database_key_index,
-                error.to,
-                query_stack.iter().map(|query| query.database_key_index),
-                &mut cycle,
+                to_id,
+                |aq| v.push(aq.database_key_index),
             );
-            cycle.push(database_key_index);
-
-            assert!(!cycle.is_empty());
-
-            for active_query in query_stack
-                .iter_mut()
-                .filter(|query| cycle.iter().any(|key| *key == query.database_key_index))
-            {
-                active_query.cycle = cycle.clone();
+            Cycle::new(Arc::new(v))
+        };
+        debug!("cycle {:?}", cycle.debug(db));
+
+        // Identify cycle recovery strategy:
+        let recovery_strategy = self.mutual_cycle_recovery_strategy(db, &cycle);
+        debug!("cycle recovery strategy {:?}", recovery_strategy);
+
+        // If using fallback, we have to mark the cycle participants, so they know to recover.
+        match recovery_strategy {
+            CycleRecoveryStrategy::Panic => {}
+            CycleRecoveryStrategy::Fallback => {
+                // Mark the cycle participants, so they know to recover:
+                dg.for_each_cycle_participant(
+                    from_id,
+                    &mut from_stack,
+                    database_key_index,
+                    to_id,
+                    |aq| {
+                        aq.cycle = Some(cycle.clone());
```

Comment:
Here, I'd expect us to also set `durability` and `changed_at`.

Reply:
Let me see if I can write a test for this...

Reply:
Yup, here's the test: https://github.com/nikomatsakis/salsa/pull/3 :) Really like that the new logic is straightforward -- I am able to just read and understand the code now, without spending quite some time to just get back into salsa.

Reply:
Nice. Let me check that out!

Reply:
Is this the test I was wanting earlier? Quite possible =)

Reply:
OK, I've pushed a fix for this test, and we now compute the changed-at and durability across all cycle participants more cleanly. I am reasonably happy with this state, though I am not *thrilled* that we invoke the recover function for C0 twice and from within probe. I want to create some tests for what happens when the recover function itself induces a cycle. =)